models/agent_model.go

```diff
@@ -442,13 +442,14 @@ func (s *Agent) DSN(service *Service, dialTimeout time.Duration, database string
 		}
 		switch {
 		case password != "":
-			u.User = url.UserPassword(username, password)
+			u.User = url.UserPassword(username, "__password__")
```

Comment:
> Why just not to use `url.UserPassword(username, url.QueryEscape(password))`

Just see https://gitlab.com/go-yp/mongodb-escape-password-research/:
| SOLUTION | TOTAL | SUCCESS |
|----------|-------|---------|
| 0        | 36    | 38      |
| 1        | 10    | 38      |
| 2        | 35    | 38      |
| 3        | 36    | 38      |
| 4        | 37    | 38      |
| 5        | 1     | 38      |

Generate 38 password with specials symbols:
```go
	var symbols = "" +
		";,/?:@&=+$" +
		"-_.!~*'()" +
		"#" +
		" " +
		"%" +
		"[]" +
		"абвгґдеє" +
		"\""

	// on "go run main.go" FAILED ONLY "'"

	for i, symbol := range symbols {
		password := fmt.Sprintf("example%spassword", string(symbol))

		// password = strings.Replace(password, "'", "''", -1)

		userPasswords = append(userPasswords, [3]string{
			"tester" + strconv.Itoa(i),
			password,
			"/tester_db",
		})
	}
```

test solutions:
```go
func ping(username, password, db string, solution int) error {
	u := &url.URL{
		Scheme:   "mongodb",
		Host:     "localhost:27017",
		Path:     db,
		RawQuery: "connectTimeoutMS=2000&serverSelectionTimeoutMS=2000",
	}

	var (
		dsn     string
		setAuth = false
	)

	switch solution {
	// SOLUTION NOW EXISTS IN pmm-manager
	// 36 / 38
	case solutionExists:
		u.User = url.UserPassword(username, password)

		dsn = u.String()
	// Artem's propose solution https://github.com/percona/pmm-managed/pull/1098#discussion_r861838935
	// 10 / 38
	case solutionArtem:
		u.User = url.UserPassword(username, url.QueryEscape(password))

		dsn = u.String()
	// Another solution https://github.com/percona/pmm-managed/pull/1003/files
	// 35 / 38
	case solutionPR:
		const plusPlaceholder = "__plus__" // There is a risk of error, when the password contains this string. Can change it to a more unique string.

		password = strings.ReplaceAll(password, "+", plusPlaceholder)

		u.User = url.UserPassword(username, password)

		dsn = u.String()

		dsn = strings.ReplaceAll(dsn, url.QueryEscape(plusPlaceholder), "%2B")
	// 36 / 38
	case solutionMy1:
		u.User = url.UserPassword(username, "__password__")

		dsn = u.String()

		dsn = strings.Replace(dsn, "__password__", url.QueryEscape(password), 1)
	// 37 / 38
	case solutionMy2:
		u.User = url.UserPassword(username, "__password__")

		dsn = u.String()

		dsn = strings.Replace(dsn, ":__password__@", ":"+url.QueryEscape(password)+"@", 1)
	//  1 / 38
	case solutionMy3:
		dsn = u.String()

		setAuth = true
	}

	ctx := context.Background()

	opts := options.Client().ApplyURI(dsn)
	if setAuth {
		opts = opts.SetAuth(options.Credential{
			Username: username,
			Password: password,
		})
	}

	client, err := mongo.Connect(ctx, opts)
	if err != nil {
		return err
	}
	defer client.Disconnect(ctx) //nolint:errcheck

	if err = client.Ping(ctx, nil); err != nil {
		return err
	}

	return nil
}
```

So main problem that in one sender side standard Go package "url" use one logic for stringify password to valid URL, in receiver side go.mongodb.org/mongo-driver/mongo use another logic for escape this password from URL.

In standard Go we have:
```go
// String returns the encoded userinfo information in the standard form
// of "username[:password]".
func (u *Userinfo) String() string {
	if u == nil {
		return ""
	}
	s := escape(u.username, encodeUserPassword)
	if u.passwordSet {
		s += ":" + escape(u.password, encodeUserPassword)
	}
	return s
}
```
```go
// Package url parses URLs and implements query escaping.
package url

escape(u.password, encodeUserPassword)
```

In go.mongodb.org/mongo-driver/mongo :
```go
p.Password, err = url.QueryUnescape(password)
```
```
// Package url parses URLs and implements query escaping.
package url

func QueryUnescape(s string) (string, error) {
	return unescape(s, encodeQueryComponent)
}
```

Reply:
Damn. Also it seems that we are not alone with this problem: https://kecci.medium.com/handle-mongo-connection-with-percent-encoding-in-go-c8b52b58094f

But what is more important that `net/url` is not suitable for MongoDB URIs. Maybe we can use something that mongo driver uses instead of `net/url` here?

Reply:
MongoDB driver use "go.mongodb.org/mongo-driver/x/mongo/driver/connstring", but connstring work only with parsing https://pkg.go.dev/go.mongodb.org/mongo-driver/x/mongo/driver/connstring

ConnString.String() not stringify URL
```go
type ConnString struct {
	Original                           string
	// ...
}

func (p *parser) parse(original string) error {
	p.Original = original
	// ...
	return nil
}

func (u *ConnString) String() string {
	return u.Original
}
```


> Maybe we can use something that mongo driver uses instead of `net/url` here?
No, ConnString cannot do it, and i do search in MongoDB driver by *.go "Password:"

Reply:
@YaroslavPodorvanov @artemgavrilov 
I will try to have [this](https://jira.percona.com/browse/PMM-9947)  done soon. It should make all secure and fix this kind of problems. Because it will be encrypted and in base64 (e.g. no encoding issues)

Reply:
@YaroslavPodorvanov @artemgavrilov please take a [look](https://github.com/percona/pmm-agent/pull/372)